src/runtime/slot-composer.ts

```diff
@@ -48,81 +37,32 @@ export class SlotComposer {
-    // options.rootContainer = options.rootContainer || options.rootContext || (options.containers || Object).root;
-    // assert((options.rootContainer !== undefined)
-    //        !==
-    //        (options.noRoot === true),
-    //   'Root container is mandatory unless it is explicitly skipped');
-    // this._containerKind = options.containerKind;
-    // if (options.modalityName) {
-    //   this.modality = Modality.create([options.modalityName]);
-    // }
-    this.modalityHandler = opts.modalityHandler;
-
-    //if (opts.noRoot) {
-    //  return;
-    //}
-
-    const containerByName =
-        opts.containers
-        //|| this.findRootContainers(options.rootContainer)
-        //|| {}
-        ;
-    //if (Object.keys(containerByName).length === 0) {
-      // fallback to single 'root' slot using the rootContainer.
-      //containerByName['root'] = opts.rootContainer;
-    //}
-
-    Object.keys(containerByName).forEach(slotName => {
-      this._contexts.push(ProvidedSlotContext.createContextForContainer(
-        `rootslotid-${slotName}`, slotName, containerByName[slotName], [`${slotName}`]));
+    // See SlotUtils::findAllSlotCandidates
+    Object.keys(opts.containers).forEach(slotName => {
+      const context = this.createContextForContainer(slotName);
+      this._contexts.push(context);
     });
   }
 
-  // findRootContainers(rootContainer) {
-  //   return this.modalityHandler.slotConsumerClass.findRootContainers(rootContainer)
-  // }
-
-  get consumers(): SlotConsumer[] {
-    return this._consumers;
-  }
-  get containerKind(): string {
-    return this._containerKind;
-  }
-
-  getSlotConsumer(particle: Particle, slotName: string): SlotConsumer {
-    return this.consumers.find(s => s.consumeConn.particle === particle && s.consumeConn.name === slotName);
-  }
-
-  findContainerByName(name: string): HTMLElement | undefined  {
-    // const contexts = this.findContextsByName(name);
-    // if (contexts.length === 0) {
-    //   // TODO this is a no-op, but throwing here breaks tests
-    //   console.warn(`No containers for '${name}'`);
-    // } else if (contexts.length === 1) {
-    //   return contexts[0].container;
-    // }
-    // console.warn(`Ambiguous containers for '${name}'`);
-    return undefined;
-  }
-
-  // TODO(sjmiles): only returns ProvidedSlotContexts, why is it called 'findContexts'?
-  findContextsByName(name: string): ProvidedSlotContext[] {
-    const filter = ctx => (ctx instanceof ProvidedSlotContext) && (ctx.name === name);
-    return this._contexts.filter(filter) as ProvidedSlotContext[];
+  createContextForContainer(name) {
+    return {
+      id: `rootslotid-${name}`,
+      name,
+      tags: [`${name}`],
+      spec: new ProvideSlotConnectionSpec({name}),
+      handles: []
+    };
   }
 
-  findContextById(slotId: string) {
-    return this._contexts.find(({id}) => id === slotId) || {};
+  getAvailableContexts() {
+    return this._contexts.concat(this.arc.activeRecipe.slots);
```

Comment:
three dots deep...

Reply:
Yes, it's not ideal, but there was no good information architecture to begin with. My tack is to perform this massive cleanup work as a first step, and then circle back to clean up the remaining ugly bits. It's too hard to fix everything all at once.

Reply:
Fwiw, note also there is no return type, that's because the downstream code uses various different types to refer to this data (and uses type-casting to avoid TS complaints). In other words, it started out in bad shape, and I plan to improve it more it moving forward. SlotContext itself is gone.